### ISSUE TITLE
[FR] VSCode - Some way to edit the layout string

### ISSUE THREAD

**Issue by aceslick911** [statelyai/editor-feedback#77]
_This is a feature request based on a bug I’m having being productive with the editor. Until I have some time to contribute useful debug info for the bug I just wanted to suggest ways around / a fix:_

- Create an option to store the @xstate-layout string as an external json file.
- Currently, if I make manual edits to a statechart (such as moving a state between machines or levels) the entire custom layout is lost and an auto layout is applied which usually requires 5/10 mins of re-arranging for large Statecharts. 
- If we could view/edit the layout data, I would be able to make edits in the code- view without having to spend time re-arranging the layout 🥹

**Comment by aceslick911** [comment]
Also - if anyone has any tips on how to decode the layout string this would reealy help please 🙏

**Comment by davidkpiano** [comment]
Added to Linear:
- https://linear.app/statelyai/issue/STA-1302/extension-move-layout-string-to-external-file
- https://linear.app/statelyai/issue/STA-1303/extension-preserve-manual-layout-when-syncing-with-code

**Comment by mattpocock** [comment]
Some things that could be improved here:

1. Saving the xstate-layout string to an external xstate.lock file makes sense, we can put some time into this when we have some.

2. Currently, we need the ENTIRE info about the graph when we perform an autolayout, and the info is linked to the names of the various states. When you change a state's name or rearrange it in the graph (move it to a parent or something) it busts the entire layout info and we need to layout everything again.

One way to do this would be to implement partial information autolayout, which could use the layout info we have for most pieces of the graph and just autolayout the rest.

Another way to do it would be to adjust the layout info while you edit. This covers the use case you're running into above, but it also means that everyone editing the layout info MUST be using the VSCode extension, and you can't make edits from GitHub without running into the same issue as above.

Probably a combination of both approaches is the way to go.

**Comment by void-mAlex** [comment]
just adding a quick thought here, on the external state, please make it toggle-able config, I really prefer it be together with the machine, but I can see the argument(s) for it living outside.

what I've been doing to get around the auto layout issue (though may not work in all cases), if you have the editor open at the same time as changing the machine directly in the file, you can generally hit the undo button.
what I found happens quite often is it restores the layout of the states that it knows about, but anything that was moved, clumps up at 0,0 .

this means I now only have a handful of states to re-arrange. in practice this works remarkably well.

this behaviour has the benefit of already being partially there. at the very least I'm proposing this as a quick in between solution.

in my opinion automatic autolayout should only happen if you don't have any layout information

**Comment by davidkpiano** [comment]
> in my opinion automatic autolayout should only happen if you don't have any layout information

I agree with this. I'm currently implementing a POC for autolayout for individual nodes and edges when they are added. This will replace overeager autolayout and provide sensible default positioning for new entities.

**Comment by aceslick911** [comment]
Sorry I posted this and didn’t come back to it for a month 😅

appreciate the work being done here and holy cow I’m going to try that workaround thanks @void-mAlex !!
